File: 16-06-2025_Equalize_the_Towers.java
class Solution {
    public long getTotalCost(int[] heights, int[] cost, int target) {
        long total = 0;
        for (int i = 0; i < heights.length; i++) {
            total += 1L * Math.abs(heights[i] - target) * cost[i];
        }
        return total;
    }

    public int minCost(int[] heights, int[] cost) {
        int low = Integer.MAX_VALUE;
        int high = Integer.MIN_VALUE;
        
        // Find min and max heights
        for (int h : heights) {
            low = Math.min(low, h);
            high = Math.max(high, h);
        }

        long answer = getTotalCost(heights, cost, heights[0]);

        // Binary search on target height
        while (low < high) {
            int mid = (low + high) / 2;
            long cost1 = getTotalCost(heights, cost, mid);
            long cost2 = getTotalCost(heights, cost, mid + 1);
            answer = Math.min(cost1, cost2);

            if (cost1 < cost2) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return (int) getTotalCost(heights, cost, low);
    }
}

File: 16-06-2025_Equalize_The_Towers.py
class Solution:
    def minCost(self, heights, cost):
        def get_total_cost(target):
            total = 0
            for h, c in zip(heights, cost):
                total += abs(h - target) * c
            return total

        low = min(heights)
        high = max(heights)

        # Binary search on the target height
        while low < high:
            mid = (low + high) // 2
            cost1 = get_total_cost(mid)
            cost2 = get_total_cost(mid + 1)

            if cost1 < cost2:
                high = mid
            else:
                low = mid + 1

        return get_total_cost(low)
